rollup/internal/controller/watcher/l1_watcher.go

```diff
@@ -116,52 +117,39 @@ func (w *L1WatcherClient) SetConfirmations(confirmations rpc.BlockNumber) {
 // FetchBlockHeader pull latest L1 blocks and save in DB
 func (w *L1WatcherClient) FetchBlockHeader(blockHeight uint64) error {
 	w.metrics.l1WatcherFetchBlockHeaderTotal.Inc()
-	fromBlock := int64(w.processedBlockHeight) + 1
-	toBlock := int64(blockHeight)
-	if toBlock < fromBlock {
-		return nil
+
+	var block *gethTypes.Header
+	block, err := w.client.HeaderByNumber(w.ctx, big.NewInt(int64(blockHeight)))
+	if err != nil {
+		log.Warn("Failed to get block", "height", blockHeight, "err", err)
+		return err
 	}
-	if toBlock > fromBlock+contractEventsBlocksFetchLimit {
-		toBlock = fromBlock + contractEventsBlocksFetchLimit - 1
+
+	if block == nil {
+		log.Warn("Received nil block", "height", blockHeight)
+		return errors.New("received nil block")
 	}
 
-	var blocks []orm.L1Block
-	var err error
-	height := fromBlock
-	for ; height <= toBlock; height++ {
-		var block *gethTypes.Header
-		block, err = w.client.HeaderByNumber(w.ctx, big.NewInt(height))
-		if err != nil {
-			log.Warn("Failed to get block", "height", height, "err", err)
-			break
-		}
-		var baseFee uint64
-		if block.BaseFee != nil {
-			baseFee = block.BaseFee.Uint64()
-		}
-		blocks = append(blocks, orm.L1Block{
-			Number:          uint64(height),
-			Hash:            block.Hash().String(),
-			BaseFee:         baseFee,
-			GasOracleStatus: int16(types.GasOraclePending),
-		})
+	var baseFee uint64
+	if block.BaseFee != nil {
+		baseFee = block.BaseFee.Uint64()
 	}
 
-	// failed at first block, return with the error
-	if height == fromBlock {
-		return err
+	l1Block := orm.L1Block{
+		Number:          blockHeight,
+		Hash:            block.Hash().String(),
+		BaseFee:         baseFee,
+		GasOracleStatus: int16(types.GasOraclePending),
 	}
-	toBlock = height - 1
 
-	// insert succeed blocks
-	err = w.l1BlockOrm.InsertL1Blocks(w.ctx, blocks)
+	err = w.l1BlockOrm.InsertL1Blocks(w.ctx, []orm.L1Block{l1Block})
 	if err != nil {
-		log.Warn("Failed to insert L1 block to db", "fromBlock", fromBlock, "toBlock", toBlock, "err", err)
+		log.Warn("Failed to insert L1 block to db", "blockHeight", blockHeight, "err", err)
```

Comment:
So if there's a reorg and we fetch the same block twice, we'll fail with this warning here? (If yes then that's OK)

Reply:
Yes. I think based on this it would insert failed:
```
create unique index l1_block_hash_uindex
on l1_block (hash) where deleted_at IS NULL;

create unique index l1_block_number_uindex
on l1_block (number) where deleted_at IS NULL;
```